src/lib/lockfile.ts

```diff
@@ -0,0 +1,111 @@
+import * as fs from 'fs';
+import { dirname } from 'path';
+import { TypedError } from 'typed-error';
+
+import { exec, unlinkAll } from './fs-utils';
+
+// Equivalent to `drwxrwxrwt`
+const STICKY_WRITE_PERMISSIONS = 0o1777;
+export const NOBODY_UID = 65534;
+
+/**
+ * Internal lockfile manager to track files in memory
+ */
+// Track locksTaken, so that the proper locks can be cleaned up on process exit
+const locksTaken: { [lockName: string]: boolean } = {};
+
+// Returns all current locks taken, as they've been stored in-memory.
+export const getLocksTaken = (): string[] => Object.keys(locksTaken);
+
+// Try to clean up any existing locks when the process exits
+process.on('exit', () => {
+	for (const lockName of getLocksTaken()) {
+		try {
+			unlockSync(lockName);
+		} catch (e) {
+			// Ignore unlocking errors
+		}
+	}
+});
+
+interface ChildProcessError {
+	code: number;
+	stderr: string;
+	stdout: string;
+}
+
+export class LockfileExistsError extends TypedError
+	implements ChildProcessError {
+	public code: number;
+	public stderr: string;
+	public stdout: string;
+
+	constructor(path: string) {
+		super();
+		this.code = 73;
+		this.stderr = `lockfile: Sorry, giving up on "${path}"`;
+		this.stdout = '';
+	}
+}
+
+export async function lock(path: string, uid = NOBODY_UID) {
+	/**
+	 * Set parent directory permissions to `drwxrwxrwt` (octal 1777), which are needed
+	 * for lockfile binary to run successfully as the `nobody` (UID 65534) user.
+	 * Otherwise the `nobody` user will not have write permissions to the necessary
+	 * bind-mounted lockfile directories. `chmod` does not fail or throw if the
+	 * directory already has the proper permissions.
+	 */
+	await fs.promises.chmod(dirname(path), STICKY_WRITE_PERMISSIONS);
+
+	// Run the lockfile binary as UID 65534. See https://linux.die.net/man/1/lockfile
+	// `-r 0` means that lockfile will not retry if the lock exists.
+	try {
+		// Lock the file using binary
+		await exec(`lockfile -r 0 ${path}`, { uid });
+		// Store a lock in memory as taken
+		locksTaken[path] = true;
+	} catch (error) {
+		// Code 73 refers to EX_CANTCREAT (73) in sysexits.h, or:
+		// A (user specified) output file cannot be created.
+		// See: https://nxmnpg.lemoda.net/3/sysexits
+		if (error instanceof LockfileExistsError) {
+			// If error code is 73, updates.lock file already exists.
+			// Throw this error directly where it's parsed into a Supervisor-specific
```

Comment:
Is not really parsed anymore right.

Reply:
Yeah, the service name and app ID that are included in the lockfile error message are in  the scope of the parent function, so no longer need to be parsed from the error message thrown here. Updated this in the next force push.